### DISCUSSION TITLE
Replace docker with orbstack

### DISCUSSION THREAD

**Discussion by arendarenko** [wardenenv/warden#714] [Ideas]
Hi guys!

[OrbStack](https://orbstack.dev/) seems to be good alternative to docker in terms of speed and CPU usage. Considering that it [supports](https://docs.orbstack.dev/compare/docker-desktop) most of the docker features, it would be great [at least as an option] to use warden with it.

**Comment by navarr** [comment]
OrbStack appears to support Docker commands.  I don't _currently_ have the bandwidth to test it, but there's no immediate reason for me to believe Warden wouldn't work.  If you'd like to test-run it, and report what issues you find and you're comfortable editing shell scripts, I'd be happy to lend my knowledge of Warden in the form of advice to what might need modifying to make it work.

**Reply by navarr** [reply]
Just want to note here that I've been using Orbstack as my daily driver on Mac for quite some time with no issues - it should be plug and play!

**Reply by bap14** [reply]
I've been running it as my only docker environment for the last 18 months or so. Never had an issue.

**Comment by sszymkowiczpg** [comment]
I've been using OrbStack with Warden for a few weeks now and I can confirm it works flawlessly :)

**Comment by bap14** [comment]
@navarr @arendarenko I've been using orbstack for months now with Warden, and it's fine. Even the migration from Docker Desktop volume data to Orbstack data kept projects in the same state. Since it's a drop-in replacement for Docker, everything in Warden "just works". What Orbstack does is expose some additional avenues to connect using the `orb` command, but the normal `warden` commands work just fine (from my usage).

**Reply by bap14** [reply]
@arendarenko @navarr That being said, I would _not_ recommend "replacing" docker with Orbstack. Orbstack is MacOS only, and Warden does currently support Windows and Linux. So "replacing" docker with Orbstack would mean we'd be limited to MacOS only.

That said, @arendarenko you can submit a PR for the docs to make mention of Orbstack as an alternative to Docker Desktop on MacOS (though technically I think any Docker replacement would work, so long as it supports the "docker" command(s), e.g. lima/colima).

**Reply by navarr** [reply]
Good callout - though I would never remove Docker support from Warden

**Comment by arendarenko** [comment]
I'm surprised how active warden community is :) 

I've read all your replies and tried Orbstack - it just works! 

For those who are going to try, the action list is below:

- [Install](https://docs.orbstack.dev/quick-start#installation) Orbstack
- Run the app
- Click on "Migrate Docker Data" button (note: it will **clone** existing docker images/containers/volumes, so make sure you have enough space on disk)
- Run warden global services 
- Open your project and run the environment, same way as you did it with Docker

--

One thing I'm going to check further (and maybe someone of you already checked?): do we still need mutagen to be enabled or not.

**Reply by TuVanDev** [reply]
@arendarenko You still need to use Mutagen.
Note: Avoid starting your project containers through the OrbStack app (User interface) because it won't create the sync session in Mutagen. Instead, run the `warden env start` command to start an environment as normal.

**Comment by MagePsycho** [comment]
Is there a proper guideline on moving from Docker Desktop to Orbstack with Warden?
So that I don't lose any DB volumes etc.

Thanks

**Reply by bap14** [reply]
@MagePsycho Install OrbStack, then open the OrbStack application and in the menu select "File --> Migrate Docker Data". It will move all existing Docker volumes to OrbStack. Then just start your environments and no data should be lost. It worked fine for me.

<img width="289" alt="image" src="https://github.com/wardenenv/warden/assets/15036826/637f4a6a-0690-45e3-918c-c94c8228098e">

**Comment by joshfedo** [comment]
If anyone runs into "Cannot connect to the Docker daemon at unix:///Users/[user-name]/.docker/run/docker.sock. Is the docker daemon running?" you may need to create a symlink:

sudo ln -s /Users/[user-name]/.orbstack/run/docker.sock /Users/[user-name]/.docker/run/docker.sock

**Reply by nkapila6** [reply]
Lifesaver!

**Comment by hardyjohnson** [comment]
Maybe this could be a discussion about generalizing the code base and supporting a bring your own container orchestration binary while defining an interface that is supported (docker compose version > x), vs replace with another specific tool (orbstack, colima, LXC, etc.)?

**Comment by erikhansen** [comment]
I finally made the switch to OrbStack. I migrated from Docker Desktop using the "File > Migrate from Docker Desktop…" option and after sorting out the issue below, all seems to be working well. Thanks for everyone's work on this—looking forward to my M1 Max 64GB machine not running into memory usage issues going forward.

I'm using:

* MacOS 15.4
* Warden 0.15.0
* OrbStack 2.0.1

I ran into this error:

```
warden env up
Connecting traefik to examplesite_default network
Connecting tunnel to examplesite_default network
Connecting mailhog to examplesite_default network
[+] Running 8/8
 ✔ Container examplesite-php-fpm-1           Started     1.2s 
 ✔ Container examplesite-php-debug-1         Started     0.9s 
 ✔ Container examplesite-db-1                Started     0.1s 
 ✔ Container examplesite-elasticsearch-hq-1  Started     0.1s 
 ✔ Container examplesite-rabbitmq-1          Started     0.2s 
 ✔ Container examplesite-elasticsearch-1     Started     0.2s 
 ✔ Container examplesite-redis-1             Started     0.2s 
 ✔ Container examplesite-nginx-1             Started     0.2s 
Started Mutagen daemon in background (terminate with "mutagen daemon stop")     
Connecting to agent (POSIX)...                                                  
Error: unable to connect to beta: unable to connect to endpoint: unable to dial agent endpoint: unable to classify agent handshake error: unknown process exit error
ERROR: Command `mutagen sync create -c "${MUTAGEN_SYNC_FILE}" --label "warden-sync=${WARDEN_ENV_NAME}" --ignore "${WARDEN_SYNC_IGNORE:-}" "${WARDEN_ENV_PATH}${WARDEN_WEB_ROOT:-}" "docker://$($WARDEN_BIN env ps -q php-fpm)/var/www/html"` at /opt/homebrew/Cellar/warden/0.14.3/commands/sync.cmd:53 failed with exit code 1
```

I worked around the issue above by following the steps in [this Github issue](https://github.com/mutagen-io/mutagen/issues/531):

```
brew upgrade mutagen

    ==> Upgrading havoc-io/mutagen/mutagen
      0.12.0 -> 0.18.1

mutagen daemon stop
mutagen daemon start
warden sync start
```
